modules/auxiliary/server/capture/ldap.rb

```diff
@@ -87,142 +82,37 @@ def on_dispatch_request(client, data)
-                    state[client][:domain] = domains.empty? ? nil : domains.join('.')
-                  rescue InvalidDNError => e
-                    print_error("LDAP Login Attempt => From:#{state[client][:name]} DN:#{user_login.name}")
-                    raise e
-                  end
-                elsif user_login.name =~ /\\/
-                  pub_info = user_login.name.split('\\')
-                  if pub_info.length <= 2
-                    state[client][:user] = pub_info[1]
-                    state[client][:domain] = pub_info[0]
-                  else
-                    result_code = Net::LDAP::ResultCodeInvalidCredentials
-                    print_error("LDAP Login Attempt => From:#{state[client][:name]} DN:#{user_login.name}")
-                  end
-                else
-                  state[client][:user] = user_login.name
-                  state[client][:domain] = nil
-                  result_code = Net::LDAP::ResultCodeInvalidCredentials
-                end
-                state[client][:private] = user_login.authentication
-                state[client][:private_type] = :password
-                unless state[client][:user].empty? && state[client][:private].empty?
-                  report_cred(state[client])
-                end
-                result_message = "LDAP Login Attempt => From:#{state[client][:name]} Username:#{state[client][:user]} Password:#{state[client][:pass]}"
-                result_message += " Domain:#{state[client][:domain]}" if state[client][:domain]
-                print_good(result_message)
-              elsif user_login.authentication[0] == 'GSS-SPNEGO'
-                if user_login.authentication[1] =~ /NTLMSSP/
-                  message = user_login.authentication[1]
-
-                  if message[8, 1] == "\x01"
-                    domain = datastore['Domain']
-                    server = datastore['Server'] # parse the domain and everythingfrom the type 1 received
-                    dnsname = datastore['DnsName']
-                    dnsdomain = datastore['DnsDomain']
-                    dom, ws = parse_type1_domain(message)
-                    if dom
-                      domain = dom
-                    end
-                    if ws
-                      server = ws
-                    end
-                    mess1 = Rex::Text.encode_base64(message)
-                    hsh = MESSAGE.process_type1_message(mess1, @challenge, domain, server, dnsname, dnsdomain)
-                    chalhash = Rex::Text.decode_base64(hsh)
-                    response = encode_ldapsasl_response(
+              
+              auth_info = process_login_request(user_login)
+              auth_info = auth_info.merge(state[:client])
+              if auth_info[:error_msg]
+                print_error(auth_info[:error_msg])
+              else
+                if user_login.authentication[1] =~ /NTLMSSP/ && auth_info[:ntlm_t2]
+                  response = encode_ldapsasl_response(
```

Comment:
This looks reasonable for now, I suspect if another `authentication` pattern was added required extension of `process_login_request()` this would not be the best way to test the result and take action. By implementing the next stage of auth here in the `MetasploitModule` vs delegating it to a common `Server::Auth` method the `authentication` related code gets fragmented making it more difficult to add specs for the end to end `auth` process and increasing the maintenance burden.

Consider adding a method of detecting further authentication requirements are needed based on the return value from `process_login_request` and passing the request to a continuation method. Looking at the code path already here is looks like `:result_code` already provides a possible solution:

``` Ruby
                if auth_info[:result_code] == Net::LDAP::ResultCodeSaslBindInProgress
                  return Server::Auth.continued_sasl_auth(pdu, auth_info)
                else
```

Also the I believe the proposed `continued_sasl_auth()` here could be the complete `res` needed for this client request input so you would not need perform the `on_send_response` in this portion of the conditional and could leave the send to the default behavior for handling `res` when set to a value.